### ISSUE TITLE
start_polling(), no clean way to exit? (etc)

### ISSUE THREAD

**Issue by zefoo** [python-telegram-bot/python-telegram-bot#209]
A few questions numbered for easy reference:

1. If I am using virtualenv, and I'm activated in the environment, and if I run my code via python -m myteleapp (code nearly identical to https://github.com/python-telegram-bot/python-telegram-bot/blob/master/examples/echobot2.py ) ... which starts polling, I can't seem to exit out of polling with ctrl+c. Only way is to kill -9 my app and I don't think this is good for a shutdown. Is there any other way to break out of start_polling()? Perhaps I am doing something wrong. 

2. Or, is there a way to shutdown start_polling if I send the bot a specific message (which I would valid against a certain chat_id)? I tried sys.exit(0) in echo-like function from here https://github.com/python-telegram-bot/python-telegram-bot/blob/master/examples/echobot2.py

3. Does start_polling() poll every second? Or what's the delay? It looks like default is 0 from the code. Does that mean that it's immediate? This is partly why I went with webhooks, I thought it would be quicker, but this may be my misunderstanding.

4. The reason I'm wanting to break out of start_polling() is because if I import another mod from my echo-like module, and if I update this other module that's included, I want to make sure start_polling() calls functions with the latest code. I believe I have to do this, I don't believe start_polling() would account for updates to code alone (since the process would stay active).

Thank you.

**Comment by jh0ker** [comment]
Hello @zefoo 

1. That sounds like you forgot to call `updater.idle()` after `updater.start_polling()`. 

2. You should stop the bot via `updater.stop()`

3. There are two parameters for `start_polling()`, `poll_interval` with a default of 0 and `timeout` with a default of 10. This is called "long-polling". The bot connects to the telegram server and waits 10 seconds. If an update is received in this time, the connection is closed immediately, the update gets processed and after waiting for `poll_interval`, a new one is opened. If there is no update after 10 seconds, the same happens without any updates.

4. I'm not sure if I understand you, but maybe it would be enough to de-register the old handlers from the dispatcher and register the new ones. This can be done while it's running.

**Comment by zefoo** [comment]
Thank you. My questions may express some confusion, I'm trying to wrap my head around this (and I'm fairly new to python). (It may help to quickly https://bpaste.net/+python something, that might help me wrap my head around this)

1. I was using this nearly verbatim: https://github.com/python-telegram-bot/python-telegram-bot/blob/master/examples/echobot2.py - I had updater.idle(). It appears to work if I call my code with python -c 'from myapp import main; main()' - then I can press ctrl+c and it shuts down. If I do: python -m myapp it doesn't close with ctrl+c. Not sure if I'm doing something wrong here? 

2. I'm slightly confused, how does this get called if the code runs start_polling() and continues until cancelled? I tried to put updater.stop() in a message handler, but the scope for updater didn't roll over into the handler, so that didn't work. Maybe I run updater.stop() from a different module at a different time?

3. Thank you, I assume when the bot connects to the telegram server it's a persistent connection for 10 seconds (eg. not checking every 1 second for updates, but consistently connected). 

4. Per my confusion on #2, how do I de-register old handlers, exactly? 

5. Another issue I'm having is, similar to #4, how do I get a new session for sqlalchemy every time a new message is received? Maybe I start/close sqlalchemy session in a message handler. I didn't have to think about many of these things with webhook.

Thank you jh0ker, appreciate it!

**Comment by jh0ker** [comment]
1. If I run `python -m echobot2`, I am able to stop the bot with Ctrl+C, so you must have changed something. Maybe you could share a minimal working example so I can take a look.

2. You would want to call `updater.stop()` from within the handler. You have to make the `updater` variable globally accessible. Check Google for more information about variable scoping in Python, but it's not that different from other programming languages.

3. It is indeed a persistent connection. Check [Wikipedia](https://en.wikipedia.org/wiki/Push_technology#Long_Polling) for more info.

4. You can use `Dispatcher.removeTelegramCommandHandler` and similar. There is one remove method for each add method. Check the `stable` documentation [here](http://python-telegram-bot.readthedocs.org/en/stable/py-modindex.html) for more info (we're currently preparing for v4.0, so the `latest` version might have inaccurate information). You will probably have to make the `updater` variable globally scoped for this as well.

**Comment by zefoo** [comment]
1. Thank you for your willingness. I tested with exact echobot2 and you're right, there is something with my code. Taking things outside of main() so updater is global fixed this (which is ok for now).

2. Thought so, but it doesn't seem to work as expected. I have this: https://bpaste.net/show/c1aeb0630893 - I send the bot 'stop', and it says "Stopping", and then I see this:

2016-03-20 15:22:45,004 - telegram.updater - INFO - Stopping Updater and Dispatcher... 

And it seems to hang here, the process never finishes. Any idea what might cause this? (It's been this way for over several minutes now). This might be the issue I'm noticing in #1, too (except I disabled logging so I was missing that, lesson learned).

3. Aha, great! 

4. So, per my pastebin above, if I understand correctly, I can do Dispatcher.removeTelegramCommandHandler within the echo() function as long as updater is global scope. Great. 

5. I ended up using a context dict that defines the session for sqlalchemy. I rewrote my code to support the context dict and everything works great now (the context dict gets passed throughout the code), and I create a new context in every new message received)

Only remaining item is #2 and I'll close this. Maybe updater.idle() isn't suppose to end the process. Thank you jh0ker.

**Comment by jh0ker** [comment]
We had a change in threading management that seems to cause a deadlock if you were to call `updater.stop()` inside a handler. Maybe you could use the `signal` module to send a SIGINT or SIGABRT to the process, this should be taken care of asynchronically and should unblock the `updater.idle()` call

**Comment by zefoo** [comment]
Thank you. I've tried sending every signal listed prior to posting this and it still seems to hang on something like "2016-03-20 15:22:45,004 - telegram.updater - INFO - Stopping Updater and Dispatcher...". If this is simply a bug, we can close this now.

**Comment by jh0ker** [comment]
Pinging @tsnoam about that thing, apart from that I'll close this now as you suggested

**Comment by rustcat** [comment]
3. There are two parameters for `start_polling()`, `poll_interval` with a default of 0 and `timeout` with a default of 10. This is called "long-polling". The bot connects to the telegram server and waits 10 seconds. If an update is received in this time, the connection is closed immediately, the update gets processed and after waiting for `poll_interval`, a new one is opened. If there is no update after 10 seconds, the same happens without any updates.

(Hope I'm not violating protocol by responding to this question 3+ years later).  My bot finally works as designed on my machine (yay!) and I'm planning to move it to pythonanywhere.   I'm based in the US.  Should I be concerned that my bot may timeout sometimes?  What steps if any should I take on update bot polling to ensure it is up all the time for users to interact with?  Thank you.

**Comment by Poolitzer** [comment]
Hey, what do you mean with being concerned that it timeouts?

Your second question "To ensure its up all the time" is not related to the issue (as far as I can tell), but as long as the updater can make a connection to telegram, it will stay up all the time.
